### PR TITLE
Adds combined pledge move effects

### LINKED ISSUE #289: MOVE_WATER_PLEDGE Effect

### LINKED ISSUE #290: MOVE_FIRE_PLEDGE Effect

### LINKED ISSUE #291: MOVE_GRASS_PLEDGE Effect

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Closes #291
Closes #290
Closes #289

Credit to Skeli and Lunos. The idea is largely from the CFRU and a lot of BattleScript code is from Lunos.

### CODE REVIEW

File: data/battle_scripts_1.s [hunk 1]
```diff
@@ -439,6 +439,97 @@ gBattleScriptsForMoveEffects::
 	.4byte BattleScript_EffectHit                     @ EFFECT_GIGATON_HAMMER
 	.4byte BattleScript_EffectSaltCure                @ EFFECT_SALT_CURE
 
+BattleScript_EffectPledge::
+	attackcanceler
+	accuracycheck BattleScript_PrintMoveMissed, ACC_CURR_MOVE
```
Comment (Skeli789) [inline data/battle_scripts_1.s:444]: There shouldn't be an accuracy check for the first Pokemon (the one that waits for its partner).

File: data/battle_scripts_1.s [hunk 2]
```diff
@@ -439,6 +439,97 @@ gBattleScriptsForMoveEffects::
...
+
+BattleScript_TheSwampDisappeared::
+	printstring STRINGID_THESWAMPDISAPPEARED
+	waitmessage B_WAIT_TIME_LONG
+	end2
+
+BattleScript_EffectHit_Pledge::
+	attackcanceler
+	accuracycheck BattleScript_PrintMoveMissed, ACC_CURR_MOVE
+	printstring STRINGID_THETWOMOVESBECOMEONE
```
Comment (Skeli789) [inline data/battle_scripts_1.s:512]: This should be before the accuracy check

File: data/battle_scripts_1.s [hunk 3]
```diff
@@ -439,6 +439,97 @@ gBattleScriptsForMoveEffects::
...
+	waitanimation
+	end2
+
+BattleScript_TheSwampDisappeared::
+	printstring STRINGID_THESWAMPDISAPPEARED
+	waitmessage B_WAIT_TIME_LONG
+	end2
+
+BattleScript_EffectHit_Pledge::
+	attackcanceler
```
Comment (Skeli789) [inline data/battle_scripts_1.s:510]: You've called the attackcancler twice? Once when the move starts, and another when it redirects to an actual attack.

File: data/battle_scripts_1.s [hunk 4]
```diff
@@ -439,6 +439,97 @@ gBattleScriptsForMoveEffects::
...
+	accuracycheck BattleScript_PrintMoveMissed, ACC_CURR_MOVE
+	printstring STRINGID_THETWOMOVESBECOMEONE
+	waitmessage B_WAIT_TIME_LONG
+	ppreduce
```
Comment (Skeli789) [inline data/battle_scripts_1.s:514]: Same with ppreduce, there are two now.

File: src/battle_script_commands.c [hunk 5]
```diff
@@ -8444,6 +8446,9 @@ static bool32 CourtChangeSwapSideStatuses(void)
     UPDATE_COURTCHANGED_BATTLER(tailwindBattlerId);
     UPDATE_COURTCHANGED_BATTLER(luckyChantBattlerId);
     UPDATE_COURTCHANGED_BATTLER(stickyWebBattlerId);
+    UPDATE_COURTCHANGED_BATTLER(rainbowBattlerId);
+    UPDATE_COURTCHANGED_BATTLER(seaOfFireBattlerId);
+    UPDATE_COURTCHANGED_BATTLER(swampBattlerId);
```
Comment (Skeli789) [inline src/battle_script_commands.c:8451]: You don't need battler ids for these effects as they persist irrespective of the Pokemon that used them. You can swap the timers with `COURTCHANGE_SWAP`.

File: src/battle_script_commands.c [hunk 6]
```diff
@@ -16320,3 +16325,106 @@ void BS_ApplySaltCure(void)
...
+            gBattlescriptCurrInstr = BattleScript_EffectCombinedPledge_Fire;
+        }
+        else if ((gCurrentMove == MOVE_WATER_PLEDGE && partnerMove == MOVE_FIRE_PLEDGE)
+              || (gCurrentMove == MOVE_FIRE_PLEDGE && partnerMove == MOVE_WATER_PLEDGE))
+        {
+            gCurrentMove = MOVE_WATER_PLEDGE;
+            gBattlescriptCurrInstr = BattleScript_EffectCombinedPledge_Water;
+        }
+    }
+    else if ((gChosenActionByBattler[partner] != B_ACTION_USE_ITEM || gChosenActionByBattler[partner] != B_ACTION_SWITCH)
```
Comment (Skeli789) [inline src/battle_script_commands.c:16359]:
```suggestion
    else if (gChosenActionByBattler[partner] == B_ACTION_USE_MOVE
```

File: src/battle_util.c [hunk 7]
```diff
@@ -11167,6 +11254,8 @@ u32 CalcSecondaryEffectChance(u8 battler, u8 secondaryEffectChance)
 {
     if (GetBattlerAbility(battler) == ABILITY_SERENE_GRACE)
         secondaryEffectChance *= 2;
+    if (gSideTimers[GetBattlerSide(battler)].rainbowTimer && gCurrentMove != MOVE_SECRET_POWER)
+        secondaryEffectChance *= 2;
```
Comment (Skeli789) [inline src/battle_util.c:11258]: Two problems with this.
1. I think this is missing from the CFRU, but apparently, according to Bulbapedia, the rainbow can't double flinch chances. You'll need to modify this to take in a move and not just the secondaryEffectChance. 
2. You can't call `gCurrentMove` in this util function. It's not guaranteed it will be called at attack time. For example, the AI could use it in its calls. Instead, change the parameter to a move (like in the first suggestion) and check that instead.

**Review by Skeli789 — CHANGES_REQUESTED** [review]

**Comment by Skeli789** [comment]
Two more things:
1. If the expansion currently supports the Gen 8 change where the turn order is recalculated before each attack, make sure this does not happen while `pledgeMove` is set.
2. I didn't see anything that changes the power to `150` when the pledge moves have been combined?

File: src/battle_script_commands.c (hunk 5 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c:8451]: I can't use `COURTCHANGE_SWAP` since there is no status flag. Would using `SWAP` be enough?

**Comment by AlexOn1ine** [comment]
> Two more things:
> 
>     1. If the expansion currently supports the Gen 8 change where the turn order is recalculated before each attack, make sure this does not happen while `pledgeMove` is set.

Yeah, it does have it. 
> 
>     2. I didn't see anything that changes the power to `150` when the pledge moves have been combined?

Forgot.

File: src/battle_util.c (hunk 7 above)
Comment (AlexOn1ine) [reply src/battle_util.c:11258]: > the rainbow can't double flinch chances
Isn't it only in cases with Serene Grace?

File: test/battle/move_effect/pledge.c [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        MESSAGE("The opposing Foe Wynaut was hurt by the sea of fire!");
+        ANIMATION(ANIM_TYPE_STATUS, B_ANIM_STATUS_BRN, opponentLeft);
+        MESSAGE("The opposing Foe Wobbuffet was hurt by the sea of fire!");
+        ANIMATION(ANIM_TYPE_STATUS, B_ANIM_STATUS_BRN, opponentRight);
+        MESSAGE("The opposing Foe Wynaut was hurt by the sea of fire!");
+        MESSAGE("The sea of fire around the opposing team disappeared!");
+    }
+}
+
+DOUBLE_BATTLE_TEST("f Fire deals 1/8th damage per turn")
```
Comment (sphericalice) [inline test/battle/move_effect/pledge.c:98]: I think you're missing "Sea O" here

File: src/battle_script_commands.c (hunk 5 above)
Comment (Skeli789) [reply src/battle_script_commands.c:8451]: Might be good to add a side status flag then, as these are side statuses after all.

File: src/battle_util.c (hunk 7 above)
Comment (Skeli789) [reply src/battle_util.c:11258]: You're right, I misread. Seems CFRU is doing it correctly then. No change is needed then other than making a specific case for flinching.

File: test/battle/hold_effect/gems.c [hunk 9]
```diff
@@ -87,3 +87,17 @@ SINGLE_BATTLE_TEST("Gem is consumed if the move type is changed")
...
+        PLAYER(SPECIES_DELCATTY) { Item(ITEM_CHOPLE_BERRY); };
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN {
+            MOVE(opponent, MOVE_ROCK_SMASH);
+        }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_ROCK_SMASH, opponent);
+    }
+}
```
Comment (LOuroboros) [inline test/battle/hold_effect/gems.c:103]: Please make sure the file ends with an empty line :eyes:

Comment (AlexOn1ine) [reply test/battle/hold_effect/gems.c]: This test shouldn't have been there :grimacing:

File: test/battle/move_effect/pledge.c (hunk 8 above)
Comment (AlexOn1ine) [reply test/battle/move_effect/pledge.c]: fixed

File: src/battle_script_commands.c (hunk 5 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c]: Added a status flag

**Comment by AlexOn1ine** [comment]
> If the expansion currently supports the Gen 8 change where the turn order is recalculated before each attack, make sure this does not happen while pledgeMove is set.

It seems like the expansion doesn't have dynamic speed.

@Skeli789 I, hopefully, fixed all issues.

**Comment by LOuroboros** [comment]
Are all those changes related to `CalcSecondaryEffectChance` truly needed?

**Comment by AlexOn1ine** [comment]
> Are all those changes related to `CalcSecondaryEffectChance` truly needed?

For the rainbow effect to work? Yes.
I know it isn't pretty but I wasn't able to come up with something better. I especially didn't like the Triple Arrows situation because the move has 2 effects and one isn't boosted by serene grace if rainbow is up.

**Comment by LOuroboros** [comment]
> For the rainbow effect to work? Yes.

I mean, that's not really true; it *can* work without the change just fine by using `gCurrentMove`.

I just read the review Skeli wrote above though, and I can with them in that it may not be always be reliable. The change is certainly not pretty, but I guess that can't be helped, so carry on 👍

File: src/battle_util.c (hunk 7 above)
Comment (AlexOn1ine) [reply src/battle_util.c:11311]: I changed the parameter to a move but would it be fine to keep `secondaryEffectChance` and keep checking the current move if a different function was made to account for the AI calls?
e.g.
```
CanBeParalyzed
AI_CanBeParalyzed
```

File: data/battle_scripts_1.s (hunk 4 above)
Comment (AlexOn1ine) [reply data/battle_scripts_1.s:514]: should be fixes

File: data/battle_scripts_1.s (hunk 3 above)
Comment (AlexOn1ine) [reply data/battle_scripts_1.s]: should be fixed.

File: data/battle_scripts_1.s (hunk 2 above)
Comment (AlexOn1ine) [reply data/battle_scripts_1.s:511]: should be fixed.

File: src/battle_script_commands.c (hunk 6 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c]: fixed

File: data/maps/SlateportCity_PokemonCenter_1F/map.json [hunk 10]
```diff
@@ -12,7 +12,7 @@
   "allow_running": false,
   "show_map_name": false,
   "battle_scene": "MAP_BATTLE_SCENE_NORMAL",
-  "connections": null,
```
Comment (Skeli789) [inline data/maps/SlateportCity_PokemonCenter_1F/map.json]: These map file changes seem unrelated to the PR?

File: include/config.h [hunk 11]
```diff
@@ -6,7 +6,7 @@
 // still has them in the ROM. This is because the developers forgot
 // to define NDEBUG before release, however this has been changed as
 // Ruby's actual debug build does not use the AGBPrint features.
-#define NDEBUG
+// #define NDEBUG
```
Comment (Skeli789) [inline include/config.h]: Is this change intentional?

File: include/constants/battle.h [hunk 12]
```diff
@@ -219,10 +219,10 @@
 // Per-side statuses that affect an entire party
 #define SIDE_STATUS_REFLECT                 (1 << 0)
 #define SIDE_STATUS_LIGHTSCREEN             (1 << 1)
-#define SIDE_STATUS_STICKY_WEB              (1 << 2)
+#define SIDE_STATUS_STICKY_WEB              (1 << 2 | 1 << 3)
```
Comment (Skeli789) [inline include/constants/battle.h:223]: What's the rational for using two status bits for this and future attacks?

File: src/battle_script_commands.c [hunk 13]
```diff
@@ -3668,8 +3668,15 @@ void SetMoveEffect(bool32 primary, u32 certain)
...
             case MOVE_EFFECT_TRIPLE_ARROWS:
                 {
-                    u8 randomLowerDefenseChance = RandomPercentage(RNG_TRIPLE_ARROWS_DEFENSE_DOWN, CalcSecondaryEffectChance(gBattlerAttacker, 50));
-                    u8 randomFlinchChance = RandomPercentage(RNG_TRIPLE_ARROWS_FLINCH, CalcSecondaryEffectChance(gBattlerAttacker, 30));
+                    u8 randomLowerDefenseChance;
+                    u8 randomFlinchChance;
+
+                    randomLowerDefenseChance = RandomPercentage(RNG_TRIPLE_ARROWS_DEFENSE_DOWN, 50 * CalcSecondaryEffectChance(gBattlerAttacker, gCurrentMove));
+
+                    if (GetBattlerAbility(gBattlerAttacker) == ABILITY_SERENE_GRACE && gSideStatuses[GetBattlerSide(gBattlerAttacker)] & SIDE_STATUS_RAINBOW)
```
Comment (Skeli789) [inline src/battle_script_commands.c:3675]:
```suggestion
                    if (GetBattlerAbility(gBattlerAttacker) == ABILITY_SERENE_GRACE || (gSideStatuses[GetBattlerSide(gBattlerAttacker)] & SIDE_STATUS_RAINBOW))
```

File: src/battle_util.c [hunk 14]
```diff
@@ -2501,16 +2501,25 @@ u8 DoFieldEndTurnEffects(void)
...
-                {
-                    BattleScriptExecute(BattleScript_RainbowContinues);
-                    effect++;
+                    for (gBattlerAttacker = 0; gBattlerAttacker < gBattlersCount; gBattlerAttacker++)
+                    {
+                        if (GetBattlerSide(gBattlerAttacker) == side)
+                            break;
+                    }
+
+                    if (--gSideTimers[side].rainbowTimer == 0)
```
Comment (Skeli789) [inline src/battle_util.c:2512]: You're missing a check for `gSideTimers[side].rainbowTimer > 0`. Otherwise if the status flag is set and the timer is 0 it'll cause underflow. Same for the other pledge statuses.

File: src/battle_util.c [hunk 15]
```diff
@@ -8628,7 +8657,11 @@ static u16 CalcMoveBasePower(u16 move, u8 battlerAtk, u8 battlerDef)
     switch (gBattleMoves[move].effect)
     {
     case EFFECT_PLEDGE:
-        // todo
+        // if (gBattleStruct->pledgeMove)
```
Comment (Skeli789) [inline src/battle_util.c]: Remove the commented out code here.

File: src/battle_util.c [hunk 16]
```diff
@@ -11250,11 +11283,18 @@ bool8 AreBattlersOfOppositeGender(u8 battler1, u8 battler2)
...
 
-u32 CalcSecondaryEffectChance(u8 battler, u8 secondaryEffectChance)
+u32 CalcSecondaryEffectChance(u8 battler, u16 move)
 {
-    if (GetBattlerAbility(battler) == ABILITY_SERENE_GRACE)
+    u8 secondaryEffectChance = gBattleMoves[move].secondaryEffectChance;
+    bool8 hasSereneGrace = (GetBattlerAbility(battler) == ABILITY_SERENE_GRACE);
+
+    if (gSideStatuses[GetBattlerSide(battler)] & SIDE_STATUS_RAINBOW && hasSereneGrace && gBattleMoves[move].effect == EFFECT_FLINCH_HIT)
+        return secondaryEffectChance;
```
Comment (Skeli789) [inline src/battle_util.c:11303]: As we discussed earlier, the chance for flinching is still x2 for either, it's just not x4 when both are active.

File: src/battle_util.c [hunk 17]
```diff
@@ -11250,11 +11283,18 @@ bool8 AreBattlersOfOppositeGender(u8 battler1, u8 battler2)
     return (gender1 != MON_GENDERLESS && gender2 != MON_GENDERLESS && gender1 != gender2);
 }
 
-u32 CalcSecondaryEffectChance(u8 battler, u8 secondaryEffectChance)
+u32 CalcSecondaryEffectChance(u8 battler, u16 move)
 {
-    if (GetBattlerAbility(battler) == ABILITY_SERENE_GRACE)
+    u8 secondaryEffectChance = gBattleMoves[move].secondaryEffectChance;
+    bool8 hasSereneGrace = (GetBattlerAbility(battler) == ABILITY_SERENE_GRACE);
```
Comment (Skeli789) [inline src/battle_util.c:11299]: If you're going to make a variable for `hasSereneGrace`, would be cleaner if you made a variable `hasRainbow = (gSideStatuses[GetBattlerSide(battler)] & SIDE_STATUS_RAINBOW) != 0` also.

File: src/battle_util.c [hunk 18]
```diff
@@ -11250,11 +11283,18 @@ bool8 AreBattlersOfOppositeGender(u8 battler1, u8 battler2)
     return (gender1 != MON_GENDERLESS && gender2 != MON_GENDERLESS && gender1 != gender2);
 }
 
-u32 CalcSecondaryEffectChance(u8 battler, u8 secondaryEffectChance)
```
Comment (Skeli789) [inline src/battle_util.c]: In order to avoid the current workaround for Triple Arrows, you can change the function to
`CalcSecondaryEffectChance(u8 battler, u8 secondaryEffectChance, u16 moveEffect)`

That way for Triple Arrows, you could call this function once for the effect of lowering defense, and again for the effect of flinching.

File: src/battle_script_commands.c [hunk 19]
```diff
@@ -16404,25 +16416,24 @@ void BS_SetPledge(void)
 
 void BS_SetPledgeStatus(void)
```
Comment (Skeli789) [inline src/battle_script_commands.c:16438]: Is there no check here for if the status is already set? As far as I'm aware, the timer doesn't reset if the effect is already active (unless I'm misremembering).

**Review by Skeli789 — CHANGES_REQUESTED** [review]

File: data/maps/SlateportCity_PokemonCenter_1F/map.json (hunk 10 above)
Comment (AlexOn1ine) [reply data/maps/SlateportCity_PokemonCenter_1F/map.json]: Yep, sorry. These were caused by opening porymap and were removed in a later commit.

File: include/config.h (hunk 11 above)
Comment (AlexOn1ine) [reply include/config.h]: No, forgot to revert it back. 

Fixed in a later commit.

File: src/battle_util.c (hunk 15 above)
Comment (AlexOn1ine) [reply src/battle_util.c]: fixed

File: src/battle_util.c (hunk 17 above)
Comment (AlexOn1ine) [reply src/battle_util.c:11299]: Yeah, this is much better. Will do.

File: src/battle_script_commands.c (hunk 19 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c:16438]: Am I understanding it correctly that a combined pledge move attack still happens but the timer isn't reset?

File: src/battle_util.c (hunk 18 above)
Comment (AlexOn1ine) [reply src/battle_util.c]:
```
RandomPercentage(RNG_TRIPLE_ARROWS_DEFENSE_DOWN, CalcSecondaryEffectChance(gBattlerAttacker, 50, moveEffect));
RandomPercentage(RNG_TRIPLE_ARROWS_FLINCH, CalcSecondaryEffectChance(gBattlerAttacker, 30, EFFECT_FLINCH_HIT));

```
Just to make sure I understood you correctly. I made the suggested change and turned the Triple Arrow workaround into the previous 2 calls while adding `EFFECT_FLINCH_HIT` to the flinch check. Would this be correct?

File: src/battle_util.c (hunk 16 above)
Comment (AlexOn1ine) [reply src/battle_util.c:11303]: Yep, I missed this one.

File: src/battle_util.c (hunk 14 above)
Comment (AlexOn1ine) [reply src/battle_util.c:2512]: How can an underflow happen if the status flag is removed once the timer is at 0?
```
if (--gSideTimers[side].rainbowTimer == 0)
{
    gSideStatuses[side] &= ~SIDE_STATUS_RAINBOW;
    BattleScriptExecute(BattleScript_TheRainbowDisappeared);
    effect++;
}
```

File: include/constants/battle.h (hunk 12 above)
Comment (AlexOn1ine) [reply include/constants/battle.h:223]: Maybe someone more knowledgeable then me can comment on why it needs 2 bits but as far as I can tell the change in this PR is only visually.

Comment (Skeli789) [reply include/constants/battle.h:223]: What do you mean? Sticky Web before was just using one flag `0x4`, but you changed it to use two flags `0x4` and `0x8`. Yes, ultimately it functions identically, but there's now an extra bit that's being wasted.

File: src/battle_util.c (hunk 14 above)
Comment (Skeli789) [reply src/battle_util.c:2512]: Say the user modifies the repo to sets the status flag at the start of the battle (eg. for a field effect). The current code would decrease the timer from 0 to `0xFF` and after 255 turns remove the field effect.

File: src/battle_util.c (hunk 18 above)
Comment (Skeli789) [reply src/battle_util.c]: It would work, but for clarity pass in `EFFECT_DEFENSE_DOWN_HIT` explicitly instead of `moveEffect` on the first line

File: src/battle_script_commands.c (hunk 19 above)
Comment (Skeli789) [reply src/battle_script_commands.c:16438]: You can confirm with an ingame test, but generally when a field effect is set the timer isn't reset to maximum value each time the attack is used.

File: include/constants/battle.h (hunk 12 above)
Comment (LOuroboros) [reply include/constants/battle.h:223]: This change came from my never submitted PR for the combined moves system + pledge moves, where I made the change for readability reasons after noticing that some of the `SIDE_STATUS` flags were skipping a value.

For example, after `SIDE_STATUS_STICKY_WEB` which has a value of 0x4, what's defined is a `SIDE_STATUS_SPIKES` with a value of 0x10, effectively skipping a `SIDE_STATUS` 0x8 flag entirely.

When I locally made it so all the flags were defined one after the next *(1 << 0, 1 << 1, 1 << 2, 1 << 3, 1 << 4, 1 << 5 and so on)*, I noticed that stuff broke until I reverted it, so I just assumed that all the values that were skipped were actually being used but not defined explicitly.

Alex just copied that change as is.

File: src/battle_util.c (hunk 14 above)
Comment (AlexOn1ine) [reply src/battle_util.c:2512]: I see, this makes a lot of sense.

File: include/constants/battle.h (hunk 12 above)
Comment (Skeli789) [reply include/constants/battle.h:223]: Ah, I can say for certain then those values certainly aren’t being used for Sticky Web. Using both bits will likely cause whatever issues you discovered before. You can leave a comment saying that the 0x8 bit for example is used for some unknown flag and shouldn’t be used.

Comment (AlexOn1ine) [reply include/constants/battle.h:223]: > What do you mean? Sticky Web before was just using one flag 0x4, but you changed it to use two flags 0x4 and 0x8. Yes, ultimately it functions identically, but there's now an extra bit that's being wasted.

Yes, I see. 

Since I didn't have a full understanding of it, I shouldn't have add those bits. That's my bad.

File: src/battle_util.c (hunk 18 above)
Comment (AlexOn1ine) [reply src/battle_util.c]: Thx, it looks much cleaner now.

File: src/battle_util.c (hunk 7 above)
Comment (AlexOn1ine) [reply src/battle_util.c:11307]: applied

File: src/battle_script_commands.c (hunk 19 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c:16431]: Confirmed on showdown and applied change.

File: src/battle_script_commands.c (hunk 13 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c]: This was removed thanks to the suggestion.

**Comment by AlexOn1ine** [comment]
Hopefully I didn't miss anything but ready for re-review!

File: src/battle_script_commands.c [hunk 20]
```diff
@@ -2111,7 +2111,9 @@ static void Cmd_adjustdamage(void)
     }
     if (gSpecialStatuses[gBattlerAttacker].gemBoost
         && !(gMoveResultFlags & MOVE_RESULT_NO_EFFECT)
-        && gBattleMons[gBattlerAttacker].item)
+        && gBattleMons[gBattlerAttacker].item
+        && gBattleMoves[gCurrentMove].effect != EFFECT_PLEDGE
```
Comment (Skeli789) [inline src/battle_script_commands.c:2115]: Only if a combined move is used.

File: src/battle_util.c [hunk 21]
```diff
@@ -360,6 +360,7 @@ void HandleAction_UseMove(void)
                  || (GetBattlerAbility(battler) == ABILITY_STORM_DRAIN && moveType == TYPE_WATER))
                 && GetBattlerTurnOrderNum(battler) < var
                 && gBattleMoves[gCurrentMove].effect != EFFECT_SNIPE_SHOT
+                && gBattleMoves[gCurrentMove].effect != EFFECT_PLEDGE
```
Comment (Skeli789) [inline src/battle_util.c:363]: Storm Drain would still redirect a combined Water Pledge. This may need some testing.

File: src/battle_util.c [hunk 22]
```diff
@@ -7622,6 +7737,8 @@ u8 ItemBattleEffects(u8 caseID, u32 battler, bool32 moveTurn)
             #if B_SERENE_GRACE_BOOST >= GEN_5
                 if (ability == ABILITY_SERENE_GRACE)
                     atkHoldEffectParam *= 2;
+                if (gSideTimers[GetBattlerSide(battler)].rainbowTimer && gCurrentMove != MOVE_SECRET_POWER)
```
Comment (Skeli789) [inline src/battle_util.c:7740]: This is still checking for the Rainbow timer and not the side status.

File: src/battle_util.c [hunk 23]
```diff
@@ -8542,7 +8659,7 @@ static inline u32 CalcMoveBasePower(u32 move, u32 battlerAtk, u32 battlerDef, u3
     switch (gBattleMoves[move].effect)
     {
     case EFFECT_PLEDGE:
-        // todo
```
Comment (Skeli789) [inline src/battle_util.c:8545]: Does your damage calc code support special logic for the AI? Because right now the AI would still see a combined pledge move as two individual attacks with the regular base power. This isn't necessarily something that needs to be handled now -- you could open an issue to fix it later.

**Review by Skeli789 — CHANGES_REQUESTED** [review]
Looks good overall. Just a few things left now

File: src/battle_util.c (hunk 21 above)
Comment (AlexOn1ine) [reply src/battle_util.c:363]: On [showdown](https://replay.pokemonshowdown.com/gen9doublesubers-1963224510) it doesn't.

File: src/battle_util.c (hunk 23 above)
Comment (AlexOn1ine) [reply src/battle_util.c:8545]: Yeah, it does. Both damage calcs are done here, before the attack and during but you are correct with how I've done it now it would only do it during the attack.

Would something like this solve the issue? 
```
if ((move == MOVE_WATER_PLEDGE && HasMove(BATTLE_PARTNER(battlerAtk), MOVE_FIRE_PLEDGE))
 || (move == MOVE_FIRE_PLEDGE && HasMove(BATTLE_PARTNER(battlerAtk), MOVE_WATER_PLEDGE))
 || (move == MOVE_FIRE_PLEDGE && HasMove(BATTLE_PARTNER(battlerAtk), MOVE_GRASS_PLEDGE))
 || (move == MOVE_GRASS_PLEDGE && HasMove(BATTLE_PARTNER(battlerAtk), MOVE_FIRE_PLEDGE))
 || (move == MOVE_GRASS_PLEDGE && HasMove(BATTLE_PARTNER(battlerAtk), MOVE_WATER_PLEDGE))
 || (move == MOVE_WATER_PLEDGE && HasMove(BATTLE_PARTNER(battlerAtk), MOVE_GRASS_PLEDGE)))
    basePower = 150;
```
This does seem like a lot of checks but I'm not sure if there is a cleaner way.

File: src/battle_script_commands.c (hunk 20 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c:2115]: Not sure but both on showdown and gen5 gems didn't activate if pledge moves were used in isolation.

Comment (AsparagusEduardo) [reply src/battle_script_commands.c:2115]: Here's the proof:
melonDS_s0Z7pglWl4

Comment (Skeli789) [reply src/battle_script_commands.c:2115]: Seems I read that wrong on Bulbapedia. But I now noticed that this wasn't implemented:
> If a combined move is used, it will always receive the [same-type attack bonus](https://bulbapedia.bulbagarden.net/wiki/Same-type_attack_bonus) no matter which Pokémon is the one executing it, or what type the combined move is.

File: src/battle_util.c (hunk 23 above)
Comment (Skeli789) [reply src/battle_util.c:8545]: You wouldn't want to check if the partner has the move, more so if they've selected the move. One Pokemon will always choose its attack before the other, so you can check if the partner has actually selected an attack, and if so check if it selected a pledge move different from the one currently being checked.

Comment (AlexOn1ine) [reply src/battle_util.c:8545]: I don't think this can work in the expansion because damage is calculated each turn before any moves were chosen.

Comment (AlexOn1ine) [reply src/battle_util.c:8545]: Hm, maybe there could be an exception made for the pledge effect to recalc everything after the first move was selected.

Comment (AlexOn1ine) [reply src/battle_util.c:8496]: I wasn't able to do it. Would an issue ticket be okay? Maybe someone who understands this part of the code base better could look into it.

**Comment by AlexOn1ine** [comment]
Ready for re-review. 

Credit to Dat.H A for the status animations

**Comment by DizzyEggg** [comment]
Please fix conflicts

**Comment by DizzyEggg** [comment]
Also, could you provide a list of what's left / cannot be done atm?

**Comment by AlexOn1ine** [comment]
> Also, could you provide a list of what's left / cannot be done atm?

- Swamp animation
- Update AI calcs once the faster battler chose a pledge to signal that the move will have 150BP and gets stab. Other then that the calcs during the actual attack seem to work fine.

**Comment by AlexOn1ine** [comment]
> Please fix conflicts

solved

**Comment by Bassoonian** [comment]
Please fix conflicts and then hopefully this one can be merged in too (with relevant issues being opened for the swamp and the updated AI calcs) :)

**Comment by AlexOn1ine** [comment]
> Please fix conflicts and then hopefully this one can be merged in too (with relevant issues being opened for the swamp and the updated AI calcs) :)

Conflicts solved.

**Review by Bassoonian — APPROVED** [review]
